<a href="https://colab.research.google.com/github/CeliaDiazSanchez/UFV_visualizacion/blob/main/Ejercicios%20de%20clase/clase2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**PARTE DE CLASE:**

Datos sobre el signifiaca de los columnas de la base de datos:

- AMOUNT: volumen de ventas en euros
- COUNTRY: país donde se ha producido esa venta
- SUBBRAND:  el producto
- YEAR: año   para cuando se ha hecho la predicción
- MONTH: mes    para cuando se ha hecho la predicción
- SCENARIO: AI_forecast, una venta que se ha predicho o actual venta que se ha producido
- FORECAST: actual las columnas son null, AI_forecast, en el periodo en el que se ha hecho, periodo dos que es el mes de enero, es siempre el numero menos uno para saber el mes
- FORECAST_YEAR: el año en el que se hizo la preducción


Preguntas:
1.	Número de filas
2.	Número de actuals y forecasts
3.	Horizonte de predicción
4.	Número de países y de productos
5.	Histórico de actuals o de forecast
6.	Cuantos forecast tiene (tener en cuenta el año tamb)


Importamos los datos:

In [824]:
import pandas as pd
import matplotlib.pyplot as plt
import requests
import plotly.express as px
# URL del archivo CSV en GitHub
url = "https://raw.githubusercontent.com/CeliaDiazSanchez/UFV_visualizacion/refs/heads/main/Ejercicios%20de%20clase/datos_ejercicio_ventas.csv?token=GHSAT0AAAAAACZ6AE56KMW57TQ5UBCWKKVOZZSNYMA"
nombre = 'datos_ejercicio_ventas.csv'


# Realizar la solicitud GET para obtener el contenido del archivo
response = requests.get(url)

# Verificar si la solicitud fue exitosa (código de estado 200)
if response.status_code == 200:
     # Guardar el contenido del archivo en un archivo local
    with open(nombre, 'wb') as f:
        f.write(response.content)
    print("Archivo descargado exitosamente.")
else:
    print("Error al descargar el archivo.")

Error al descargar el archivo.


In [825]:
datos = pd.read_csv('datos_ejercicio_ventas.csv')
datos.head(10)

COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
0       Portugal         Lipton (L3)  2023     12  AI_forecast  AI_P02F   
1  Great Britain         Lipton (L3)  2023     12  AI_forecast  AI_P10F   
2          Spain      Pepsi Max (L3)  2023     12  AI_forecast  AI_P09F   
3  Great Britain            7up (L3)  2024     12  AI_forecast  AI_P10F   
4        Hungary         Lipton (L3)  2023      9  AI_forecast  AI_P03F   
5         Norway  Pepsi Regular (L3)  2025      2  AI_forecast  AI_P09F   
6        Denmark            7up (L3)  2023     12  AI_forecast    AI_PF   
7    Netherlands   Mountain Dew (L3)  2024      4  AI_forecast  AI_P02F   
8  Great Britain      Pepsi Max (L3)  2023      9  AI_forecast  AI_P02F   
9          Italy  Pepsi Regular (L3)  2023      4  AI_forecast  AI_P03F   

   FORECAST_YEAR        AMOUNT  
0         2023.0  7.543562e+05  
1         2023.0  5.600306e+05  
2         2023.0  8.850198e+04  
3         2023.0  3.632245e+05  
4         2023.0  3.961761e+05  
5         2023.0  2.667267e+04  
6         2023.0  0.000000e+00  
7         2023.0  0.000000e+00  
8         2023.0  1.138557e+07  
9         2023.0  2.074742e+06

In [826]:
datos.shape

(18666, 8)

In [827]:
num_filas = datos.shape[0]
print(f"Número de filas: {num_filas}")
num_colum = datos.shape[1]
print(f"Número de columnas: {num_colum}")

Número de filas: 18666
Número de columnas: 8


In [828]:
num_forecasts = datos[datos['SCENARIO'] == 'AI_forecast'].shape[0] #[0] para coger el número de filas
print(f"Número de forecasts: {num_forecasts}")

Número de forecasts: 17766


In [829]:
num_actual = datos[datos['SCENARIO'] == 'actual'].shape[0]
print(f"Número de actual: {num_actual}")

Número de actual: 900


Cual es el horizonte de predicción:
Tamaño de la predicción (por periodo), todas tienen el mismo horizonte

In [830]:
portugal_pepsi = datos[(datos['COUNTRY'] == 'Portugal') & (datos['SUBBRAND'] == 'Pepsi Max (L3)') & (datos['SCENARIO'] == 'AI_forecast') & (datos['FORECAST'] == 'AI_P08F')]
portugal_pepsi.head(18)

COUNTRY        SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
1510   Portugal  Pepsi Max (L3)  2024     10  AI_forecast  AI_P08F   
1857   Portugal  Pepsi Max (L3)  2024      5  AI_forecast  AI_P08F   
2228   Portugal  Pepsi Max (L3)  2024      2  AI_forecast  AI_P08F   
4863   Portugal  Pepsi Max (L3)  2024      3  AI_forecast  AI_P08F   
5023   Portugal  Pepsi Max (L3)  2023      9  AI_forecast  AI_P08F   
8536   Portugal  Pepsi Max (L3)  2024      9  AI_forecast  AI_P08F   
8751   Portugal  Pepsi Max (L3)  2023     12  AI_forecast  AI_P08F   
10065  Portugal  Pepsi Max (L3)  2023     11  AI_forecast  AI_P08F   
10139  Portugal  Pepsi Max (L3)  2024      7  AI_forecast  AI_P08F   
10183  Portugal  Pepsi Max (L3)  2025      1  AI_forecast  AI_P08F   
10241  Portugal  Pepsi Max (L3)  2024      8  AI_forecast  AI_P08F   
10262  Portugal  Pepsi Max (L3)  2024      6  AI_forecast  AI_P08F   
11753  Portugal  Pepsi Max (L3)  2024      4  AI_forecast  AI_P08F   
11849  Portugal  Pepsi Max (L3)  2024     12  AI_forecast  AI_P08F   
13169  Portugal  Pepsi Max (L3)  2024      1  AI_forecast  AI_P08F   
13235  Portugal  Pepsi Max (L3)  2024     11  AI_forecast  AI_P08F   
13988  Portugal  Pepsi Max (L3)  2023      8  AI_forecast  AI_P08F   
15832  Portugal  Pepsi Max (L3)  2023     10  AI_forecast  AI_P08F   

       FORECAST_YEAR         AMOUNT  
1510          2023.0  233111.124679  
1857          2023.0  230454.298236  
2228          2023.0  207376.207925  
4863          2023.0  218441.086609  
5023          2023.0  216684.479052  
8536          2023.0  237841.238642  
8751          2023.0  225142.744320  
10065         2023.0  213810.522910  
10139         2023.0  259488.779177  
10183         2023.0  224344.441163  
10241         2023.0  262920.230636  
10262         2023.0  251749.469196  
11753         2023.0  224350.857146  
11849         2023.0  246299.503910  
13169         2023.0  203187.681573  
13235         2023.0  234967.282500  
13988         2023.0  241763.471046  
15832         2023.0  211954.365089

In [831]:
portugal_pepsi.shape
print(f"El horizonte de predicción es: {portugal_pepsi.shape[0]}")

El horizonte de predicción es: 18


In [832]:
num_paises = datos['COUNTRY'].nunique()
num_productos = datos['SUBBRAND'].nunique()

print(f"Número de países: {num_paises}")
print(f"Número de productos: {num_productos}")

Número de países: 9
Número de productos: 6


In [833]:
combinaciones_unicas = datos[['FORECAST', 'FORECAST_YEAR']].drop_duplicates()

# Contar el número de combinaciones distintas, shape[0] coge foto el número de filas
numero_combinaciones = combinaciones_unicas.shape[0]

# Mostrar el resultado
print(f"Número de combinaciones distintas de FORECAST y FORECAST_YEAR: {numero_combinaciones}")

Número de combinaciones distintas de FORECAST y FORECAST_YEAR: 13


In [834]:
print(combinaciones_unicas)

    FORECAST  FORECAST_YEAR
0    AI_P02F         2023.0
1    AI_P10F         2023.0
2    AI_P09F         2023.0
4    AI_P03F         2023.0
6      AI_PF         2023.0
11   AI_P11F         2023.0
12   AI_P06F         2023.0
13   AI_P05F         2023.0
14   AI_P07F         2023.0
15   AI_P12F         2023.0
38   AI_P08F         2023.0
48   AI_P04F         2023.0
277      NaN            NaN


Quitamos los actual para que no salga el NaN, son los actuals

In [835]:
df = datos[datos['SCENARIO'] != 'actual']
combinaciones_unicas = df[['FORECAST', 'FORECAST_YEAR']].drop_duplicates()

# Contar el número de combinaciones distintas, shape[0] coge foto el número de filas
numero_combinaciones = combinaciones_unicas.shape[0]

# Mostrar el resultado
print(f"Número de combinaciones distintas de FORECAST y FORECAST_YEAR: {numero_combinaciones}")

Número de combinaciones distintas de FORECAST y FORECAST_YEAR: 12


In [836]:
print(combinaciones_unicas)

   FORECAST  FORECAST_YEAR
0   AI_P02F         2023.0
1   AI_P10F         2023.0
2   AI_P09F         2023.0
4   AI_P03F         2023.0
6     AI_PF         2023.0
11  AI_P11F         2023.0
12  AI_P06F         2023.0
13  AI_P05F         2023.0
14  AI_P07F         2023.0
15  AI_P12F         2023.0
38  AI_P08F         2023.0
48  AI_P04F         2023.0


Histórico de actuals y forecast:

In [837]:
# Filtrar los datos donde SCENARIO es 'forecast'
forecast = datos[datos['SCENARIO'] == 'AI_forecast']

# Ordenar los datos filtrados por YEAR y MONTH
forecast_o = forecast.sort_values(by=['YEAR', 'MONTH'])

# Obtener el primer dato "forecast"
primer_forecast = forecast_o.iloc[0]

# Mostrar el primer dato "forecast"
print("Primer dato 'forecast':")
print(primer_forecast)

# Obtener el primer dato "forecast"
ultimo_forecast = forecast_o.iloc[-1]

# Mostrar el primer dato "forecast"
print("Ultimo dato 'forecast':")
print(ultimo_forecast)

Primer dato 'forecast':
COUNTRY                   Czech
SUBBRAND         Pepsi Max (L3)
YEAR                       2023
MONTH                         1
SCENARIO            AI_forecast
FORECAST                  AI_PF
FORECAST_YEAR            2023.0
AMOUNT            317424.378124
Name: 34, dtype: object
Ultimo dato 'forecast':
COUNTRY              Hungary
SUBBRAND            7up (L3)
YEAR                    2025
MONTH                      5
SCENARIO         AI_forecast
FORECAST             AI_P12F
FORECAST_YEAR         2023.0
AMOUNT                   0.0
Name: 18543, dtype: object


In [838]:
# Filtrar los datos donde SCENARIO es 'actual'
actuals = datos[datos['SCENARIO'] == 'actual']

# Ordenar los datos filtrados por YEAR y MONTH
actuals_o = actuals.sort_values(by=['YEAR', 'MONTH'])

# Obtener el primer dato "actual"
primer_actual = actuals_o.iloc[0]

# Mostrar el primer dato "actual"
print("Primer dato 'actual':")
print(primer_actual)

# Obtener el primer dato "actual"
ultimo_actual = actuals_o.iloc[-1]

# Mostrar el primer dato "actual"
print("Ultimo dato 'actual':")
print(ultimo_actual)

Primer dato 'actual':
COUNTRY                   Spain
SUBBRAND         Pepsi Max (L3)
YEAR                       2023
MONTH                         1
SCENARIO                 actual
FORECAST                    NaN
FORECAST_YEAR               NaN
AMOUNT             83274.055497
Name: 2884, dtype: object
Ultimo dato 'actual':
COUNTRY                  Norway
SUBBRAND         Pepsi Max (L3)
YEAR                       2024
MONTH                         8
SCENARIO                 actual
FORECAST                    NaN
FORECAST_YEAR               NaN
AMOUNT           3110356.073841
Name: 17720, dtype: object


#**PRÁCTICA:**

##Limpieza de base de datos:##

Ver si tenemos datos negativos en la columna AMOUNT:

In [839]:
valores_negativos = datos[datos['AMOUNT'] < 0]

if not valores_negativos.empty:
    print("Se encontraron los siguientes valores negativos en la columna 'AMOUNT':")
    print(valores_negativos)
else:
    print("No se encontraron valores negativos en la columna 'AMOUNT'.")

Se encontraron los siguientes valores negativos en la columna 'AMOUNT':
       COUNTRY     SUBBRAND  YEAR  MONTH SCENARIO FORECAST  FORECAST_YEAR  \
6564   Hungary     7up (L3)  2023      9   actual      NaN            NaN   
7261   Hungary     7up (L3)  2024      1   actual      NaN            NaN   
7904    Norway  Lipton (L3)  2024      7   actual      NaN            NaN   
8791   Hungary     7up (L3)  2023     11   actual      NaN            NaN   
11325  Hungary     7up (L3)  2023      7   actual      NaN            NaN   
11659   Norway  Lipton (L3)  2023      9   actual      NaN            NaN   
14006  Hungary     7up (L3)  2023     12   actual      NaN            NaN   
15840   Norway  Lipton (L3)  2024      3   actual      NaN            NaN   
17174  Hungary     7up (L3)  2024      2   actual      NaN            NaN   
17719   Norway  Lipton (L3)  2024      8   actual      NaN            NaN   

              AMOUNT  
6564   -19481.652378  
7261     -188.474017  
7904    -31

Vamos a eliminar los valores negativos.

In [840]:
 datos = datos[datos['AMOUNT'] >= 0]

In [841]:
valores_negativos = datos[datos['AMOUNT'] < 0]

if not valores_negativos.empty:
    print("Se encontraron los siguientes valores negativos en la columna 'AMOUNT':")
    print(valores_negativos)
else:
    print("No se encontraron valores negativos en la columna 'AMOUNT'.")

No se encontraron valores negativos en la columna 'AMOUNT'.


Vemos si tenemos datos repetidos.

In [842]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")

conteo_duplicados_forecast = duplicados['FORECAST'].value_counts()
print("\nCantidad de duplicados por tipo de 'FORECAST':")
print(conteo_duplicados_forecast)
conteo_duplicados_actual = duplicados[duplicados['SCENARIO'] == 'actual']
print("\nCantidad de duplicados actual:")
print(conteo_duplicados_actual)

Se encontraron las siguientes filas duplicadas:
             COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
158           Norway      Pepsi Max (L3)  2024     11  AI_forecast  AI_P07F   
252      Netherlands         Lipton (L3)  2023      5  AI_forecast  AI_P03F   
272          Denmark      Pepsi Max (L3)  2025      2  AI_forecast  AI_P11F   
329          Hungary      Pepsi Max (L3)  2023     12  AI_forecast  AI_P10F   
349    Great Britain         Lipton (L3)  2024      9  AI_forecast  AI_P11F   
...              ...                 ...   ...    ...          ...      ...   
18658        Hungary      Pepsi Max (L3)  2023      4  AI_forecast  AI_P04F   
18659    Netherlands  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P02F   
18660        Denmark      Pepsi Max (L3)  2023     10  AI_forecast  AI_P10F   
18661  Great Britain  Pepsi Regular (L3)  2024      2  AI_forecast  AI_P10F   
18662        Hungary  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P07F   

   

Vemos que no tenemos actual duplicados por lo tento en los dos primeros ejercicios no vamos a tener problemas, pero por otro lado tenemos que tener cuidado en el ejericio 3 con los forecast repetidos para hacer la comparación de valores reales y los forecast.

##1. Como se distribuyen las ventas realizadas en:##
  - Cada país
  - Cada mes y año
  - Cada marca

In [843]:
# Filtrar solo las ventas realizadas (no las predicciones)
ventas_reales = datos[datos['SCENARIO'] == 'actual']
print(ventas_reales.head(10))

           COUNTRY            SUBBRAND  YEAR  MONTH SCENARIO FORECAST  \
277       Portugal      Pepsi Max (L3)  2023     10   actual      NaN   
278       Portugal            7up (L3)  2023      3   actual      NaN   
279       Portugal            7up (L3)  2023     10   actual      NaN   
280  Great Britain       7up Free (L3)  2023     10   actual      NaN   
281         Norway  Pepsi Regular (L3)  2023     10   actual      NaN   
282         Norway  Pepsi Regular (L3)  2023      3   actual      NaN   
283    Netherlands            7up (L3)  2023     10   actual      NaN   
292       Portugal         Lipton (L3)  2023     10   actual      NaN   
293  Great Britain         Lipton (L3)  2023     10   actual      NaN   
294        Hungary  Pepsi Regular (L3)  2023      3   actual      NaN   

     FORECAST_YEAR        AMOUNT  
277            NaN  1.885949e+05  
278            NaN  2.934971e+05  
279            NaN  3.484466e+05  
280            NaN  1.172553e+06  
281            NaN  3

Distribución de ventas por país:

In [844]:
import plotly.express as px

ventas_por_pais = ventas_reales.groupby('COUNTRY')['AMOUNT'].sum().reset_index()

fig_bar = px.bar(
    ventas_por_pais,
    x='COUNTRY',
    y='AMOUNT',
    title='Distribución de Ventas por País',
    labels={'COUNTRY': 'País', 'AMOUNT': 'Ventas Totales (€)'},
)
fig_bar.update_layout(xaxis=dict(tickangle=45), title_font_size=16, xaxis_title_font_size=12, yaxis_title_font_size=12)
fig_bar.show()

In [845]:
for pais in ventas_por_pais['COUNTRY'].unique():
    # Filtrar datos para el país actual
    datos_pais = ventas_reales[ventas_reales['COUNTRY'] == pais]

    # Crear el diagrama de bigotes para este país
    fig_box = px.box(
        datos_pais,
        x='AMOUNT',
        title=f'Diagrama de Bigotes de Ventas - {pais}',
        labels={'AMOUNT': 'Ventas Totales (€)'},
        color_discrete_sequence=['blue']
    )

    # Personalizar el diseño
    fig_box.update_layout(
        title_font_size=16,
        xaxis_title_font_size=12,
        yaxis_title_font_size=12,
        yaxis=dict(tickvals=[1], ticktext=[pais])
    )
    fig_box.show()

Distribución de ventas por mes:

In [846]:
# Gráfico de barras por mes y país
fig_mes_pais = px.histogram(
    ventas_reales,
    x='MONTH',
    y='AMOUNT',
    color='COUNTRY',
    title='Distribución de Ventas por Mes y País',
    labels={'MONTH': 'Mes', 'AMOUNT': 'Ventas Totales (€)', 'COUNTRY': 'País'},
    barmode='group'
)
fig_mes_pais.update_layout(
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12
)
fig_mes_pais.show()

# Gráfico de barras por mes y marca
fig_mes_marca = px.histogram(
    ventas_reales,
    x='MONTH',
    y='AMOUNT',
    color='SUBBRAND',
    title='Distribución de Ventas por Mes y Marca',
    labels={'MONTH': 'Mes', 'AMOUNT': 'Ventas Totales (€)', 'SUBBRAND': 'Marca'},
    barmode='group'
)
fig_mes_marca.update_layout(
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12
)
fig_mes_marca.show()

Distribución por mes y año:

In [847]:
# Crear la columna 'Fecha' en formato "mes/año"
ventas_por_fecha = ventas_reales.groupby(['YEAR', 'MONTH', 'COUNTRY', 'SUBBRAND'])['AMOUNT'].sum().reset_index()
ventas_por_fecha['Fecha'] = ventas_por_fecha['MONTH'].astype(str) + '/' + ventas_por_fecha['YEAR'].astype(str)

# Histograma de ventas por mes/año y país
fig_pais = px.bar(
    ventas_por_fecha,
    x='Fecha',
    y='AMOUNT',
    color='COUNTRY',
    title='Histograma de Ventas por Mes y Año (Por País)',
    labels={'Fecha': 'Mes/Año', 'AMOUNT': 'Ventas Totales (€)', 'COUNTRY': 'País'},
    barmode='group'
)
fig_pais.update_layout(
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12,
)
fig_pais.show()

# Histograma de ventas por mes/año y marca
fig_marca = px.bar(
    ventas_por_fecha,
    x='Fecha',
    y='AMOUNT',
    color='SUBBRAND',
    title='Histograma de Ventas por Mes y Año (Por Marca)',
    labels={'Fecha': 'Mes/Año', 'AMOUNT': 'Ventas Totales (€)', 'SUBBRAND': 'Marca'},
    barmode='group'
)
fig_marca.update_layout(
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12,
)
fig_marca.show()


Distribución por año:

Distribución por país:

In [848]:
# Crear un DataFrame agrupado por año, país y marca
ventas_por_año = ventas_reales.groupby(['YEAR', 'COUNTRY', 'SUBBRAND'])['AMOUNT'].sum().reset_index()
# Generar diagramas de bigotes para cada país
for pais in ventas_por_año['COUNTRY'].unique():
    datos_pais = ventas_por_año[ventas_por_año['COUNTRY'] == pais]

    fig_pais = px.box(
        datos_pais,
        x='YEAR',
        y='AMOUNT',
        title=f'Distribución de Ventas por Año - {pais}',
        labels={'YEAR': 'Año', 'AMOUNT': 'Ventas (€)'},
        color_discrete_sequence=['purple']
    )
    fig_pais.update_layout(
        title_font_size=16,
        xaxis_title_font_size=12,
        yaxis_title_font_size=12,
        xaxis=dict(tickmode='linear'),
    )
    fig_pais.show()


In [849]:
# Crear un DataFrame agrupado por año, país y ventas totales
ventas_por_año_pais = ventas_reales.groupby(['YEAR', 'COUNTRY'])['AMOUNT'].sum().reset_index()

# Generar el histograma para cada país
for pais in ventas_por_año_pais['COUNTRY'].unique():
    datos_pais = ventas_por_año_pais[ventas_por_año_pais['COUNTRY'] == pais]

    fig_pais = px.histogram(
        datos_pais,
        x='YEAR',
        y='AMOUNT',
        title=f'Distribución de Ventas por Año - {pais}',
        labels={'YEAR': 'Año', 'AMOUNT': 'Ventas (€)'},
        color='YEAR',
        barmode='group',
    )
    fig_pais.update_layout(
        title_font_size=16,
        xaxis_title_font_size=12,
        yaxis_title_font_size=12,
    )
    fig_pais.show()


Distribución por marca:

In [850]:
# Generar diagramas de bigotes para cada marca
for marca in ventas_por_año['SUBBRAND'].unique():
    datos_marca = ventas_por_año[ventas_por_año['SUBBRAND'] == marca]

    fig_marca = px.box(
        datos_marca,
        x='YEAR',
        y='AMOUNT',
        title=f'Distribución de Ventas por Año - {marca}',
        labels={'YEAR': 'Año', 'AMOUNT': 'Ventas (€)'},
        color_discrete_sequence=['green']
    )
    fig_marca.update_layout(
        title_font_size=16,
        xaxis_title_font_size=12,
        yaxis_title_font_size=12,
        xaxis=dict(tickmode='linear'),
    )
    fig_marca.show()


In [851]:
# Crear un DataFrame agrupado por año, marca y ventas totales
ventas_por_año_marca = ventas_reales.groupby(['YEAR', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

# Generar el histograma para cada marca
for marca in ventas_por_año_marca['SUBBRAND'].unique():
    datos_marca = ventas_por_año_marca[ventas_por_año_marca['SUBBRAND'] == marca]

    fig_marca = px.histogram(
        datos_marca,
        x='YEAR',
        y='AMOUNT',
        title=f'Distribución de Ventas por Año - {marca}',
        labels={'YEAR': 'Año', 'AMOUNT': 'Ventas (€)'},
        color='YEAR',
        barmode='group',
    )
    fig_marca.update_layout(
        title_font_size=16,
        xaxis_title_font_size=12,
        yaxis_title_font_size=12,
    )
    fig_marca.show()


##2. Cual es la tendencia y estacionalidad de:##
  - Todas las ventas del país con menos ventas
  - La marca con más ventas

Cogemos unicamento los datos reales, es decir en la columna SCENARIO los actial, que son las ventas realizadas.

In [852]:
# Filtrar solo las ventas realizadas (no las predicciones)
ventas_reales = datos[datos['SCENARIO'] == 'actual']
print(ventas_reales.head(10))

           COUNTRY            SUBBRAND  YEAR  MONTH SCENARIO FORECAST  \
277       Portugal      Pepsi Max (L3)  2023     10   actual      NaN   
278       Portugal            7up (L3)  2023      3   actual      NaN   
279       Portugal            7up (L3)  2023     10   actual      NaN   
280  Great Britain       7up Free (L3)  2023     10   actual      NaN   
281         Norway  Pepsi Regular (L3)  2023     10   actual      NaN   
282         Norway  Pepsi Regular (L3)  2023      3   actual      NaN   
283    Netherlands            7up (L3)  2023     10   actual      NaN   
292       Portugal         Lipton (L3)  2023     10   actual      NaN   
293  Great Britain         Lipton (L3)  2023     10   actual      NaN   
294        Hungary  Pepsi Regular (L3)  2023      3   actual      NaN   

     FORECAST_YEAR        AMOUNT  
277            NaN  1.885949e+05  
278            NaN  2.934971e+05  
279            NaN  3.484466e+05  
280            NaN  1.172553e+06  
281            NaN  3

In [853]:
ventas_pais = ventas_reales.groupby('COUNTRY')['AMOUNT'].sum().reset_index()
print("Ventas totales por país:")
print(ventas_pais)

ventas_marca = ventas_reales.groupby('SUBBRAND')['AMOUNT'].sum().reset_index()
print("\nVentas totales por marca:")
print(ventas_marca)

Ventas totales por país:
         COUNTRY        AMOUNT
0          Czech  3.535164e+07
1        Denmark  5.659668e+07
2  Great Britain  3.347786e+08
3        Hungary  4.157947e+07
4          Italy  4.345404e+07
5    Netherlands  6.395943e+07
6         Norway  5.143697e+07
7       Portugal  3.488807e+07
8          Spain  8.131266e+06

Ventas totales por marca:
             SUBBRAND        AMOUNT
0            7up (L3)  1.880808e+07
1       7up Free (L3)  3.427630e+07
2         Lipton (L3)  7.308806e+07
3   Mountain Dew (L3)  7.474391e+06
4      Pepsi Max (L3)  3.966987e+08
5  Pepsi Regular (L3)  1.398307e+08


In [854]:
pais_menos_ventas = ventas_pais.loc[ventas_pais['AMOUNT'].idxmin()]
print("\nEl país con menos ventas es:", pais_menos_ventas['COUNTRY'])
marca_mas_ventas = ventas_marca.loc[ventas_marca['AMOUNT'].idxmax()]
print("\nLa marca con más ventas es:", marca_mas_ventas['SUBBRAND'])


El país con menos ventas es: Spain

La marca con más ventas es: Pepsi Max (L3)


Una vez tenemos las datos de cual es el país con menos ventas y la marca con más ventas, vamos a hacer un diagrama de líneas para ver la tendencia y estacionalidad de ambas:


In [855]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from statsmodels.tsa.seasonal import seasonal_decompose

# Filtrar las ventas del país con menos ventas y agrupar por año y mes
ventas_espana = ventas_reales[ventas_reales['COUNTRY'] == 'Spain'].groupby(['YEAR', 'MONTH'])['AMOUNT'].sum().reset_index()

# Crear la columna 'Fecha' a partir de YEAR y MONTH
ventas_espana['Fecha'] = pd.to_datetime(ventas_espana['YEAR'].astype(str) + '-' + ventas_espana['MONTH'].astype(str))

# Establecer 'Fecha' como índice
ventas_espana.set_index('Fecha', inplace=True)
print(ventas_espana)


            YEAR  MONTH         AMOUNT
Fecha                                 
2023-01-01  2023      1  325272.938930
2023-02-01  2023      2  295588.379605
2023-03-01  2023      3  399710.273110
2023-04-01  2023      4  328035.257874
2023-05-01  2023      5  360751.156280
2023-06-01  2023      6  439198.586290
2023-07-01  2023      7  496545.236120
2023-08-01  2023      8  564336.120820
2023-09-01  2023      9  420041.980591
2023-10-01  2023     10  468425.841760
2023-11-01  2023     11  418006.199049
2023-12-01  2023     12  394147.408590
2024-01-01  2024      1  349696.533142
2024-02-01  2024      2  352899.035652
2024-03-01  2024      3  332109.256233
2024-04-01  2024      4  376678.368897
2024-05-01  2024      5  356759.011786
2024-06-01  2024      6  408967.746367
2024-07-01  2024      7  506234.362782
2024-08-01  2024      8  537862.283643


Tenemos que usar periodo un periodo más pequeño de 12 ya que solo tenemos 20 datos para hacer la tendencia y la estacionalidad y no podemos cubrir los dos periodos enteros , para ello necesitarías 24. Al tener datos de 1,5 años vamos a usar el periodo de 6(semestral).

In [856]:
# Descomponer la serie temporal (Ventas, Tendencia, Estacionalidad, Residual)
decompose_pais = seasonal_decompose(ventas_espana['AMOUNT'], model='additive', period=6)

# Crear un DataFrame para las componentes de la descomposición
df_decomposed = pd.DataFrame({
    'Fecha': ventas_espana.index,
    'AMOUNT': ventas_espana['AMOUNT'],
    'Tendencia': decompose_pais.trend,
    'Estacionalidad': decompose_pais.seasonal,
    'Residuo': decompose_pais.resid
})

# Crear la figura con plotly.express para las tres componentes
fig = px.line(df_decomposed, x='Fecha', y=['Tendencia', 'Estacionalidad', 'Residuo'],
              title=f'Descomposición de Ventas para el País con Menos Ventas: Spain',
              )

# Agregar las ventas originales a la gráfica usando plotly.graph_objects (go)
fig.add_trace(go.Scatter(x=df_decomposed['Fecha'], y=df_decomposed['AMOUNT'], mode='lines', name='Ventas',
                         line=dict(color='purple', width=2, dash='solid')))

# Configurar el diseño
fig.update_layout(
    xaxis_title="Fecha",
    yaxis_title="Ventas (€)",
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12,
    legend_title="Componentes",
    template="plotly_white"
)

# Mostrar la gráfica
fig.show()


In [857]:
# Filtrar la marca con más ventas
marca_mas_ventas = df.groupby('SUBBRAND')['AMOUNT'].sum().idxmax()

# Filtrar las ventas de la marca con más ventas y agrupar por año y mes
ventas_marca = ventas_reales[ventas_reales['SUBBRAND'] == marca_mas_ventas].groupby(['YEAR', 'MONTH'])['AMOUNT'].sum().reset_index()

# Crear la columna 'Fecha' a partir de YEAR y MONTH
ventas_marca['Fecha'] = pd.to_datetime(ventas_marca['YEAR'].astype(str) + '-' + ventas_marca['MONTH'].astype(str))

# Establecer 'Fecha' como índice
ventas_marca.set_index('Fecha', inplace=True)

print(ventas_marca)

            YEAR  MONTH        AMOUNT
Fecha                                
2023-01-01  2023      1  1.485655e+07
2023-02-01  2023      2  1.546332e+07
2023-03-01  2023      3  2.087355e+07
2023-04-01  2023      4  1.793494e+07
2023-05-01  2023      5  2.232545e+07
2023-06-01  2023      6  2.132855e+07
2023-07-01  2023      7  1.836866e+07
2023-08-01  2023      8  2.160818e+07
2023-09-01  2023      9  1.965228e+07
2023-10-01  2023     10  1.862691e+07
2023-11-01  2023     11  2.072202e+07
2023-12-01  2023     12  2.170186e+07
2024-01-01  2024      1  1.562425e+07
2024-02-01  2024      2  1.825694e+07
2024-03-01  2024      3  2.125434e+07
2024-04-01  2024      4  2.023834e+07
2024-05-01  2024      5  2.311013e+07
2024-06-01  2024      6  2.123881e+07
2024-07-01  2024      7  2.077966e+07
2024-08-01  2024      8  2.273393e+07


In [858]:
# Descomponer la serie temporal (Ventas, Tendencia, Estacionalidad, Residual)
decompose_marca = seasonal_decompose(ventas_marca['AMOUNT'], model='additive', period=6)

# Crear un DataFrame para las componentes de la descomposición
df_decomposed = pd.DataFrame({
    'Fecha': ventas_marca.index,
    'AMOUNT': ventas_marca['AMOUNT'],
    'Tendencia': decompose_marca.trend,
    'Estacionalidad': decompose_marca.seasonal,
    'Residuo': decompose_marca.resid
})

# Crear la figura con plotly.express para las tres componentes
fig = px.line(df_decomposed, x='Fecha', y=['Tendencia', 'Estacionalidad', 'Residuo'],
              title=f'Descomposición de Ventas para la Marca con Más Ventas: Pepsi Max (L3)',
            )

# Agregar las ventas originales a la gráfica usando plotly.graph_objects (go)
fig.add_trace(go.Scatter(x=df_decomposed['Fecha'], y=df_decomposed['AMOUNT'], mode='lines', name='Ventas',
                         line=dict(color='purple', width=2, dash='solid')))

# Configurar el diseño
fig.update_layout(
    xaxis_title="Fecha",
    yaxis_title="Ventas (€)",
    title_font_size=16,
    xaxis_title_font_size=12,
    yaxis_title_font_size=12,
    legend_title="Componentes",
    template="plotly_white"
)

# Mostrar la gráfica
fig.show()


### Análisis de cada gráfico:

#### 1. Gráfico de Descomposición de Ventas en el País con Menos Ventas: España

   - *Tendencia*: En el gráfico, la tendencia tiene una curva que presenta una subida durante la primera mitad de 2023, con un pico significativo en torno a julio de 2023. Posteriormente, la tendencia muestra un descenso hasta mediados de 2024, donde parece comenzar a recuperarse de nuevo.
   
   - *Estacionalidad*: La estacionalidad es bastante marcada, con fluctuaciones regulares a lo largo de cada año. Esto indica que las ventas en España presentan variaciones periódicas, probablemente relacionadas con la estacionalidad de ciertos productos o eventos en el país.
   
   - *Residuo*: Los residuos son fluctuantes, lo que sugiere que hay variaciones aleatorias en las ventas de mes a mes, algunas positivas y otras negativas. Esto puede deberse a factores externos o imprevistos que afectan a las ventas de manera no explicada por la tendencia o estacionalidad.

#### 2. Gráfico de Descomposición de Ventas para la Marca con Más Ventas: Pepsi Max

   - *Tendencia*: La tendencia aquí es bastante estable, aunque muestra un ligero aumento. Esto indica que las ventas de Pepsi Max tienen una tendencia ascendente leve, sugiriendo un crecimiento moderado a largo plazo sin grandes oscilaciones.
   
   - *Estacionalidad*: Similar al gráfico anterior, la estacionalidad es visible con picos a lo largo del tiempo. La presencia de estacionalidad muestra que las ventas de Pepsi Max también siguen un patrón periódico, lo cual puede estar asociado con la demanda cíclica del producto.
   
   - *Residuo*: El componente de residuo en este gráfico también presenta fluctuaciones, aunque parece menos pronunciado en comparación con el gráfico de España. Esto sugiere que, aunque hay variaciones mensuales en las ventas, son menos drásticas. Los residuos indican eventos o factores imprevistos que afectaron las ventas en algunos meses, pero en menor medida.


##3. Cuales son las predicciones hechas en España y como de buenas son ##

###Limpiza de FORECASTS###

Ver si tenemso filas repetidas, que solo cambie el AMOUNT, para cuando usemos las predicciones entender que pasa con los errores.

In [859]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR','AMOUNT'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")


Se encontraron las siguientes filas duplicadas:
             COUNTRY       SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
349    Great Britain    Lipton (L3)  2024      9  AI_forecast  AI_P11F   
455    Great Britain    Lipton (L3)  2024      6  AI_forecast  AI_P03F   
540    Great Britain  7up Free (L3)  2024      2  AI_forecast    AI_PF   
652    Great Britain    Lipton (L3)  2024      1  AI_forecast  AI_P05F   
789    Great Britain    Lipton (L3)  2024     11  AI_forecast  AI_P11F   
...              ...            ...   ...    ...          ...      ...   
18590        Hungary       7up (L3)  2025      1  AI_forecast  AI_P12F   
18606  Great Britain  7up Free (L3)  2024      8  AI_forecast  AI_P03F   
18631  Great Britain       7up (L3)  2024     10  AI_forecast  AI_P11F   
18649  Great Britain    Lipton (L3)  2024      5  AI_forecast  AI_P11F   
18650  Great Britain    Lipton (L3)  2024      5  AI_forecast  AI_P09F   

       FORECAST_YEAR        AMOUNT  
349           2023.0  7.36

En la base de datos tenemos 663 filas reptidas, todos sus valores, vamos a elimiar una de ellas para quitar estas filas repetidas.

In [860]:
datos = datos.drop_duplicates(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR', 'AMOUNT'])


Comprobamos que ya se han eliminado.

In [861]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR','AMOUNT'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")

No se encontraron filas duplicadas.


Ahora vamos a ver si tenemos datos duplicados en todas la columnas y que su unica variación sea AMOUNT.

In [862]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")


Se encontraron las siguientes filas duplicadas:
             COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
158           Norway      Pepsi Max (L3)  2024     11  AI_forecast  AI_P07F   
252      Netherlands         Lipton (L3)  2023      5  AI_forecast  AI_P03F   
272          Denmark      Pepsi Max (L3)  2025      2  AI_forecast  AI_P11F   
329          Hungary      Pepsi Max (L3)  2023     12  AI_forecast  AI_P10F   
380          Denmark      Pepsi Max (L3)  2024      2  AI_forecast  AI_P10F   
...              ...                 ...   ...    ...          ...      ...   
18658        Hungary      Pepsi Max (L3)  2023      4  AI_forecast  AI_P04F   
18659    Netherlands  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P02F   
18660        Denmark      Pepsi Max (L3)  2023     10  AI_forecast  AI_P10F   
18661  Great Britain  Pepsi Regular (L3)  2024      2  AI_forecast  AI_P10F   
18662        Hungary  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P07F   

   

Ver en que FORECAST tenemos esos valores repetidos.

In [863]:
conteo_duplicados_forecast = duplicados['FORECAST'].value_counts()
print("\nCantidad de duplicados por tipo de 'FORECAST':")
print(conteo_duplicados_forecast)


Cantidad de duplicados por tipo de 'FORECAST':
FORECAST
AI_P12F    768
AI_P10F    714
AI_P08F    708
AI_P11F    705
AI_P09F    690
AI_P07F    605
AI_P05F    574
AI_P04F    557
AI_P06F    550
AI_P03F    360
AI_PF      360
AI_P02F    360
Name: count, dtype: int64


  Como en este caso nos interesan los datos de españa vamos a filtar los datos por el país.

In [864]:
datos_espana = duplicados[duplicados['COUNTRY'] == 'Spain']

conteo_forecasts_espana = datos_espana['FORECAST'].value_counts()

print(conteo_forecasts_espana)

FORECAST
AI_P08F    108
AI_P04F    108
AI_P06F    108
AI_P11F    108
AI_P07F    108
AI_P12F    108
AI_P09F    108
AI_P05F    108
AI_P10F    108
Name: count, dtype: int64


Vemos que en todos tenemos valores repetidos pero en el que menos número de valores repetidos tenemos es en AI_P02F, AI_PF y AI_P03F.
Voy a dejar todos los datos pero tenemos que tener en cuenta que el error con respecto a los datos reales va a ser mayor cuento más cuantos más datos reoetidos tenemos.

Ahora los vamos a agrupar por marcas para poder comparar luego los errores por marca.

In [865]:
conteo_forecasts_por_marca = datos_espana.groupby('SUBBRAND')['FORECAST'].value_counts().unstack(fill_value=0)

print(conteo_forecasts_por_marca)

FORECAST            AI_P04F  AI_P05F  AI_P06F  AI_P07F  AI_P08F  AI_P09F  \
SUBBRAND                                                                   
7up Free (L3)            18       18       18       18       18       18   
Pepsi Max (L3)           54       54       54       54       54       54   
Pepsi Regular (L3)       36       36       36       36       36       36   

FORECAST            AI_P10F  AI_P11F  AI_P12F  
SUBBRAND                                       
7up Free (L3)            18       18       18  
Pepsi Max (L3)           54       54       54  
Pepsi Regular (L3)       36       36       36  


Estas cantidades tambien van a cuadrar con el valor del error de cada FORECAST

###Apartado resuelto sin quitar la filas que son similares que solo cambia AMOUNT###

Comparación dejando esos datos repetidos, teniendo en cuenta los resultados.

Paso 1: Filtrar las predicciones hechas en España y los datos reales
  - AI_forecast: una venta que se ha predicho - actual: venta que se ha producido



In [866]:
predicciones_spain = datos[(datos['COUNTRY'] == 'Spain') & (datos['SCENARIO'] == 'AI_forecast')]
reales_spain = datos[(datos['COUNTRY'] == 'Spain') & (datos['SCENARIO'] == 'actual')]

In [867]:
print(reales_spain)

      COUNTRY        SUBBRAND  YEAR  MONTH SCENARIO FORECAST  FORECAST_YEAR  \
296     Spain   7up Free (L3)  2023     10   actual      NaN            NaN   
414     Spain   7up Free (L3)  2024      1   actual      NaN            NaN   
728     Spain   7up Free (L3)  2024      2   actual      NaN            NaN   
729     Spain     Lipton (L3)  2024      2   actual      NaN            NaN   
835     Spain        7up (L3)  2024      4   actual      NaN            NaN   
...       ...             ...   ...    ...      ...      ...            ...   
18452   Spain   7up Free (L3)  2023      1   actual      NaN            NaN   
18467   Spain     Lipton (L3)  2023      7   actual      NaN            NaN   
18469   Spain   7up Free (L3)  2023      7   actual      NaN            NaN   
18535   Spain  Pepsi Max (L3)  2023      2   actual      NaN            NaN   
18599   Spain   7up Free (L3)  2023      9   actual      NaN            NaN   

             AMOUNT  
296    83196.895876  
414    

Primero vamos a hacer las comparaciones de españa, pero separando por marca y luego juntando todo españa, sin importar la marca.

In [868]:
import plotly.graph_objects as go

datos_espana = datos[datos['COUNTRY'] == 'Spain'].copy()

datos_actuale = datos_espana[datos_espana['SCENARIO'] == 'actual'].copy()
datos_forecaste = datos_espana[datos_espana['SCENARIO'] == 'AI_forecast'].copy()

datos_actuale['DATE'] = pd.to_datetime(datos_actuale[['YEAR', 'MONTH']].assign(DAY=1))
datos_forecaste['DATE'] = pd.to_datetime(datos_forecaste[['YEAR', 'MONTH']].assign(DAY=1))

sales_actual = datos_actuale.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

subbrands = sales_actual['SUBBRAND'].unique()

for subbrand in subbrands:
    # Filtrar los datos de marca actual
    sales_actual_subbrand = sales_actual[sales_actual['SUBBRAND'] == subbrand]

    fig = px.line(
        sales_actual_subbrand,
        x='DATE',
        y='AMOUNT',
        title=f'Ventas Reales por Submarca: {subbrand} en España',
        labels={'AMOUNT': 'Ventas Reales (€)', 'DATE': 'Fecha'},
        markers=True
    )

    # Filtrar las predicciones para esta marca
    datos_forecast_marca = datos_forecaste[datos_forecaste['SUBBRAND'] == subbrand]

    # Añadir las líneas de predicción para cada tipo de predicción (AI_P02F, AI_P03F, ...)
    for forecast_type in datos_forecast_marca['FORECAST'].unique():
        forecast_data = datos_forecast_marca[datos_forecast_marca['FORECAST'] == forecast_type]
        sales_forecast = forecast_data.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

        # Añadir las predicciones al gráfico de líneas
        fig.add_trace(
            go.Scatter(
                x=sales_forecast['DATE'],
                y=sales_forecast['AMOUNT'],
                mode='lines+markers',
                name=f'Predicción {forecast_type}',
                line=dict(dash='dash')
            )
        )

    # Configuración del diseño del gráfico
    fig.update_layout(
        title=f'Comparación de Ventas Reales y Predicciones por Marca: {subbrand} (España)',
        xaxis_title='Fecha',
        yaxis_title='Ventas (€)',
        legend_title='Escenarios',
        template="plotly_white",
        xaxis=dict(tickangle=45)
    )

    # Mostrar el gráfico
    fig.show()


Error de las predicciones de españa dividida por marca.

In [869]:
# Iterar sobre cada submarca
for subbrand in subbrands:
    # Filtrar las ventas reales para la submarca actual
    sales_actual_subbrand = sales_actual[sales_actual['SUBBRAND'] == subbrand]

    # Filtrar las predicciones para esta submarca
    df_forecast_subbrand = datos_forecaste[datos_forecaste['SUBBRAND'] == subbrand]

    # Crear lista para almacenar los errores MAE
    mae_list = []
    forecast_names = []

    # Iterar sobre cada tipo de predicción en esta submarca
    for forecast_type in df_forecast_subbrand['FORECAST'].unique():
        forecast_data = df_forecast_subbrand[df_forecast_subbrand['FORECAST'] == forecast_type]

        # Agrupar las predicciones por mes y submarca
        sales_forecast = forecast_data.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

        # Unir las ventas reales con las predicciones para esta submarca
        merged_data = pd.merge(sales_actual_subbrand, sales_forecast, on='DATE', suffixes=('_real', '_forecast'))

        # Calcular MAE
        mae = np.mean(np.abs(merged_data['AMOUNT_real'] - merged_data['AMOUNT_forecast']))
        mae_list.append(mae)

        # Añadir el nombre de la predicción
        forecast_names.append(f'{forecast_type}')

    # Crear el gráfico de barras para esta submarca
    fig_errors = go.Figure()

    # Añadir barras para el MAE
    fig_errors.add_trace(go.Bar(
        x=forecast_names,
        y=mae_list,
        name='MAE (Error Absoluto Medio)',
        marker_color='lightblue'
    ))

    # Configuración del gráfico de errores para esta submarca
    fig_errors.update_layout(
        title=f'Error MAE para {subbrand}',
        xaxis_title='Tipo de Predicción',
        yaxis_title='Error (MAE)',
        legend_title='Tipo de Error',
        template="plotly_white",
        xaxis=dict(tickangle=45)
    )

    # Mostrar el gráfico de errores para esta submarca
    fig_errors.show()


Comparación de ventas y predicciones de todas las marcas en españa juntas.

In [870]:
datos_españa = datos[datos['COUNTRY'] == 'Spain'].copy()

datos_actuals = datos_españa[datos_españa['SCENARIO'] == 'actual'].copy()
datos_forecasts = datos_españa[datos_españa['SCENARIO'] == 'AI_forecast'].copy()

datos_actuals['DATE'] = pd.to_datetime(datos_actuals[['YEAR', 'MONTH']].assign(DAY=1))
datos_forecasts['DATE'] = pd.to_datetime(datos_forecasts[['YEAR', 'MONTH']].assign(DAY=1))

sales_actual = datos_actuals.groupby('DATE')['AMOUNT'].sum().reset_index()

fig_sales = go.Figure()

fig_sales.add_trace(go.Scatter(
    x=sales_actual['DATE'],
    y=sales_actual['AMOUNT'],
    mode='lines+markers',
    name='Ventas Reales (Actual)',
    line=dict(color='blue')
))

# Añadir cada línea de predicción
for forecast_type in datos_forecasts['FORECAST'].unique():
    forecast_data = datos_forecasts[datos_forecasts['FORECAST'] == forecast_type]
    sales_forecast = forecast_data.groupby('DATE')['AMOUNT'].sum().reset_index()

    fig_sales.add_trace(go.Scatter(
        x=sales_forecast['DATE'],
        y=sales_forecast['AMOUNT'],
        mode='lines+markers',
        name=f'Predicción {forecast_type}',
        line=dict(dash='dash')
    ))

fig_sales.update_layout(
    title='Comparación de Ventas Reales vs. Predicciones España',
    xaxis_title='Fecha',
    yaxis_title='Cantidad Ventas',
    legend=dict(
        orientation="v",
        x=1.1,
        y=1,
        xanchor='left',
        yanchor='top'
    ),
    template="plotly_white"
)

fig_sales.show()


Error de españa con todas las marcas juntas.

In [871]:
# Crear listas para almacenar los errores (MAE)
mae_list = []
forecast_names = []

# Iterar sobre cada tipo de predicción
for forecast_type in datos_forecasts['FORECAST'].unique():
    forecast_data = datos_forecasts[datos_forecasts['FORECAST'] == forecast_type]

    # Agrupar las predicciones por mes
    sales_forecast = forecast_data.groupby('DATE')['AMOUNT'].sum().reset_index()

    # Unir las ventas reales con las predicciones
    merged_data = pd.merge(sales_actual, sales_forecast, on='DATE', suffixes=('_real', '_forecast'))

    # Calcular MAE
    mae = np.mean(np.abs(merged_data['AMOUNT_real'] - merged_data['AMOUNT_forecast']))
    mae_list.append(mae)

    # Añadir el nombre de la predicción
    forecast_names.append(f'{forecast_type}')

# Crear el gráfico de barras para los errores (MAE) para toda España
fig_errors = go.Figure()

# Añadir barras para el MAE
fig_errors.add_trace(go.Bar(
    x=forecast_names,
    y=mae_list,
    name='MAE (Error Absoluto Medio)',
    marker_color='blue'
))

# Configuración del gráfico de errores
fig_errors.update_layout(
    title='Error MAE para Predicciones en España',
    xaxis_title='Tipo de Predicción',
    yaxis_title='Error (MAE)',
    legend_title='Tipo de Error',
    template="plotly_white",
    xaxis=dict(tickangle=45),
    legend=dict(
        orientation="v",  # Para una leyenda vertical
        x=1.1,  # Mover la leyenda a la derecha
        y=1,  # Posición de la leyenda (ajustar según sea necesario)
        xanchor='left',  # Ancla la leyenda a la izquierda de la posición 'x'
        yanchor='top'  # Ancla la leyenda a la parte superior
    )
)

# Mostrar el gráfico de errores para toda España
fig_errors.show()


Viendo las graficas observamos que las prediciones que no tienen un error muy grande con respecto a las ventas reales son, AI_PF, AI_P02F, AI_P03F. Esto pasa porque el resto tiene repetidas y suman más.

Pero por otro lado si vemos en función de cada marca dentro de españa 7up Free (L3), Pepsi Max (L3) y Pepsi Regular (L3), siguen este mismo patrón de un error pequño en AI_PF, AI_P02F, AI_P03F. Pero por otro lado en Lipton (L3) el menos error se da en AI_P11F, AI_P12F y en 7up (L3) se da un error grnde en todos losa Forecast, con valores más o menos similares.

In [872]:
print(conteo_forecasts_espana)

FORECAST
AI_P08F    108
AI_P04F    108
AI_P06F    108
AI_P11F    108
AI_P07F    108
AI_P12F    108
AI_P09F    108
AI_P05F    108
AI_P10F    108
Name: count, dtype: int64


In [873]:
print(conteo_forecasts_por_marca)

FORECAST            AI_P04F  AI_P05F  AI_P06F  AI_P07F  AI_P08F  AI_P09F  \
SUBBRAND                                                                   
7up Free (L3)            18       18       18       18       18       18   
Pepsi Max (L3)           54       54       54       54       54       54   
Pepsi Regular (L3)       36       36       36       36       36       36   

FORECAST            AI_P10F  AI_P11F  AI_P12F  
SUBBRAND                                       
7up Free (L3)            18       18       18  
Pepsi Max (L3)           54       54       54  
Pepsi Regular (L3)       36       36       36  


Vemos que el error está en proporción a los datos repetidos calculados anteriormente, las marcas que dan datos diferentes, no tienen ningún valor reptido.

###Apartado resulto quitado las reptidas quedandonos con la media del AMOUNT en esas filas###

In [874]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")

Se encontraron las siguientes filas duplicadas:
             COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
158           Norway      Pepsi Max (L3)  2024     11  AI_forecast  AI_P07F   
252      Netherlands         Lipton (L3)  2023      5  AI_forecast  AI_P03F   
272          Denmark      Pepsi Max (L3)  2025      2  AI_forecast  AI_P11F   
329          Hungary      Pepsi Max (L3)  2023     12  AI_forecast  AI_P10F   
380          Denmark      Pepsi Max (L3)  2024      2  AI_forecast  AI_P10F   
...              ...                 ...   ...    ...          ...      ...   
18658        Hungary      Pepsi Max (L3)  2023      4  AI_forecast  AI_P04F   
18659    Netherlands  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P02F   
18660        Denmark      Pepsi Max (L3)  2023     10  AI_forecast  AI_P10F   
18661  Great Britain  Pepsi Regular (L3)  2024      2  AI_forecast  AI_P10F   
18662        Hungary  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P07F   

   

In [875]:
duplicados_espana = duplicados[duplicados['COUNTRY'] == 'Spain']
# Contar la cantidad de registros para cada 'SCENARIO' dentro de España
conteo_scenarios_espana = duplicados_espana.groupby('SCENARIO').size().reset_index(name='Cantidad')

# Mostrar el conteo
print("Cantidad de registros por SCENARIO dentro de los duplicados para España:")
print(conteo_scenarios_espana)

Cantidad de registros por SCENARIO dentro de los duplicados para España:
      SCENARIO  Cantidad
0  AI_forecast       972


Total de AI_forecast y actual en españa para ver que cuadra con el número de datos al eliminar los duplicados.

In [876]:
predicciones_spain = datos[(datos['COUNTRY'] == 'Spain') & (datos['SCENARIO'] == 'AI_forecast')]
reales_spain = datos[(datos['COUNTRY'] == 'Spain') & (datos['SCENARIO'] == 'actual')]

In [877]:
print(predicciones_spain)
print(reales_spain)

      COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
2       Spain      Pepsi Max (L3)  2023     12  AI_forecast  AI_P09F   
20      Spain  Pepsi Regular (L3)  2023     12  AI_forecast  AI_P05F   
25      Spain         Lipton (L3)  2025      3  AI_forecast  AI_P11F   
62      Spain       7up Free (L3)  2024      1  AI_forecast  AI_P04F   
68      Spain            7up (L3)  2024      4  AI_forecast  AI_P02F   
...       ...                 ...   ...    ...          ...      ...   
18585   Spain  Pepsi Regular (L3)  2024     10  AI_forecast  AI_P10F   
18611   Spain  Pepsi Regular (L3)  2024      1  AI_forecast  AI_P07F   
18617   Spain      Pepsi Max (L3)  2024     10  AI_forecast  AI_P07F   
18653   Spain  Pepsi Regular (L3)  2024      3  AI_forecast  AI_P09F   
18655   Spain  Pepsi Regular (L3)  2024     12  AI_forecast  AI_P10F   

       FORECAST_YEAR         AMOUNT  
2             2023.0   88501.980847  
20            2023.0  134268.151080  
25            2023.0 

Como tenemos 2052 datos de AI_forecast en españa y 972 son duplicados nos vamos a quedar con 1080 y actual al no tener duplicados se quedan igual.

Quitamos los duplicados poniendo en AMOUNT la media de todas las filas iguales.

In [878]:
duplicados = datos[datos.duplicated(subset=['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR'])]
if not duplicados.empty:
    print("Se encontraron las siguientes filas duplicadas:")
    print(duplicados)
else:
    print("No se encontraron filas duplicadas.")

Se encontraron las siguientes filas duplicadas:
             COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
158           Norway      Pepsi Max (L3)  2024     11  AI_forecast  AI_P07F   
252      Netherlands         Lipton (L3)  2023      5  AI_forecast  AI_P03F   
272          Denmark      Pepsi Max (L3)  2025      2  AI_forecast  AI_P11F   
329          Hungary      Pepsi Max (L3)  2023     12  AI_forecast  AI_P10F   
380          Denmark      Pepsi Max (L3)  2024      2  AI_forecast  AI_P10F   
...              ...                 ...   ...    ...          ...      ...   
18658        Hungary      Pepsi Max (L3)  2023      4  AI_forecast  AI_P04F   
18659    Netherlands  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P02F   
18660        Denmark      Pepsi Max (L3)  2023     10  AI_forecast  AI_P10F   
18661  Great Britain  Pepsi Regular (L3)  2024      2  AI_forecast  AI_P10F   
18662        Hungary  Pepsi Regular (L3)  2024      7  AI_forecast  AI_P07F   

   

In [879]:
# Separar los registros con SCENARIO = 'actual'
datos_actual = datos[datos['SCENARIO'] == 'actual']

# Separar los registros con SCENARIO distinto de 'actual'
datos_no_actual = datos[datos['SCENARIO'] != 'actual']

# Agrupar y calcular la media de 'AMOUNT' solo para los registros que no son 'actual'
datos_no_actual_agrupados = datos_no_actual.groupby(
    ['COUNTRY', 'SUBBRAND', 'YEAR', 'MONTH', 'SCENARIO', 'FORECAST', 'FORECAST_YEAR'],
    as_index=False
).agg({'AMOUNT': 'mean'})

# Combinar nuevamente los datos 'actual' sin tocar con los datos agrupados
datos_final = pd.concat([datos_actual, datos_no_actual_agrupados], ignore_index=True)

# Mostrar el resultado
print("DataFrame después de procesar:")
print(datos_final)


DataFrame después de procesar:
             COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
0           Portugal      Pepsi Max (L3)  2023     10       actual      NaN   
1           Portugal            7up (L3)  2023      3       actual      NaN   
2           Portugal            7up (L3)  2023     10       actual      NaN   
3      Great Britain       7up Free (L3)  2023     10       actual      NaN   
4             Norway  Pepsi Regular (L3)  2023     10       actual      NaN   
...              ...                 ...   ...    ...          ...      ...   
11037          Spain  Pepsi Regular (L3)  2025      3  AI_forecast  AI_P11F   
11038          Spain  Pepsi Regular (L3)  2025      3  AI_forecast  AI_P12F   
11039          Spain  Pepsi Regular (L3)  2025      4  AI_forecast  AI_P11F   
11040          Spain  Pepsi Regular (L3)  2025      4  AI_forecast  AI_P12F   
11041          Spain  Pepsi Regular (L3)  2025      5  AI_forecast  AI_P12F   

       FORECAST_YEAR

Vemos que ya tenemso una base limpia sin duplicados, volvemos a hacer el código anterios para ver en ese caso que ocurre en las prediciciones si se acercan más a los valores reales.

Paso 1: Filtrar las predicciones hechas en España y los datos reales
  - AI_forecast: una venta que se ha predicho - actual: venta que se ha producido



In [880]:
predicciones_spain = datos_final[(datos_final['COUNTRY'] == 'Spain') & (datos_final['SCENARIO'] == 'AI_forecast')]
reales_spain = datos_final[(datos_final['COUNTRY'] == 'Spain') & (datos_final['SCENARIO'] == 'actual')]

In [881]:
print(predicciones_spain)

      COUNTRY            SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
9962    Spain            7up (L3)  2023      1  AI_forecast    AI_PF   
9963    Spain            7up (L3)  2023      2  AI_forecast  AI_P02F   
9964    Spain            7up (L3)  2023      2  AI_forecast    AI_PF   
9965    Spain            7up (L3)  2023      3  AI_forecast  AI_P02F   
9966    Spain            7up (L3)  2023      3  AI_forecast  AI_P03F   
...       ...                 ...   ...    ...          ...      ...   
11037   Spain  Pepsi Regular (L3)  2025      3  AI_forecast  AI_P11F   
11038   Spain  Pepsi Regular (L3)  2025      3  AI_forecast  AI_P12F   
11039   Spain  Pepsi Regular (L3)  2025      4  AI_forecast  AI_P11F   
11040   Spain  Pepsi Regular (L3)  2025      4  AI_forecast  AI_P12F   
11041   Spain  Pepsi Regular (L3)  2025      5  AI_forecast  AI_P12F   

       FORECAST_YEAR         AMOUNT  
9962          2023.0   52156.149274  
9963          2023.0   65063.618870  
9964          2023.0 

In [882]:
print(reales_spain)

    COUNTRY        SUBBRAND  YEAR  MONTH SCENARIO FORECAST  FORECAST_YEAR  \
11    Spain   7up Free (L3)  2023     10   actual      NaN            NaN   
16    Spain   7up Free (L3)  2024      1   actual      NaN            NaN   
22    Spain   7up Free (L3)  2024      2   actual      NaN            NaN   
23    Spain     Lipton (L3)  2024      2   actual      NaN            NaN   
30    Spain        7up (L3)  2024      4   actual      NaN            NaN   
..      ...             ...   ...    ...      ...      ...            ...   
870   Spain   7up Free (L3)  2023      1   actual      NaN            NaN   
874   Spain     Lipton (L3)  2023      7   actual      NaN            NaN   
876   Spain   7up Free (L3)  2023      7   actual      NaN            NaN   
880   Spain  Pepsi Max (L3)  2023      2   actual      NaN            NaN   
889   Spain   7up Free (L3)  2023      9   actual      NaN            NaN   

           AMOUNT  
11   83196.895876  
16   60318.994231  
22   59482.8505

Quitamos los reptidos dejando una única fila con la media de los amount de las repetidas.

Primero vamos a hacer las comparaciones de españa, pero separando por marca y luego juntando todo españa, sin importar la marca.

In [883]:
import plotly.graph_objects as go

datos_espana = datos_final[datos_final['COUNTRY'] == 'Spain'].copy()

datos_actuale = datos_espana[datos_espana['SCENARIO'] == 'actual'].copy()
datos_forecaste = datos_espana[datos_espana['SCENARIO'] == 'AI_forecast'].copy()

datos_actuale['DATE'] = pd.to_datetime(datos_actuale[['YEAR', 'MONTH']].assign(DAY=1))
datos_forecaste['DATE'] = pd.to_datetime(datos_forecaste[['YEAR', 'MONTH']].assign(DAY=1))

sales_actual = datos_actuale.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

subbrands = sales_actual['SUBBRAND'].unique()

for subbrand in subbrands:
    # Filtrar los datos de marca actual
    sales_actual_subbrand = sales_actual[sales_actual['SUBBRAND'] == subbrand]

    fig = px.line(
        sales_actual_subbrand,
        x='DATE',
        y='AMOUNT',
        title=f'Ventas Reales por Submarca: {subbrand} en España',
        labels={'AMOUNT': 'Ventas Reales (€)', 'DATE': 'Fecha'},
        markers=True
    )

    # Filtrar las predicciones para esta marca
    datos_forecast_marca = datos_forecaste[datos_forecaste['SUBBRAND'] == subbrand]

    # Añadir las líneas de predicción para cada tipo de predicción (AI_P02F, AI_P03F, ...)
    for forecast_type in datos_forecast_marca['FORECAST'].unique():
        forecast_data = datos_forecast_marca[datos_forecast_marca['FORECAST'] == forecast_type]
        sales_forecast = forecast_data.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

        # Añadir las predicciones al gráfico de líneas
        fig.add_trace(
            go.Scatter(
                x=sales_forecast['DATE'],
                y=sales_forecast['AMOUNT'],
                mode='lines+markers',
                name=f'Predicción {forecast_type}',
                line=dict(dash='dash')
            )
        )

    # Configuración del diseño del gráfico
    fig.update_layout(
        title=f'Comparación de Ventas Reales y Predicciones por Marca: {subbrand} (España)',
        xaxis_title='Fecha',
        yaxis_title='Ventas (€)',
        legend_title='Escenarios',
        template="plotly_white",
        xaxis=dict(tickangle=45)
    )

    # Mostrar el gráfico
    fig.show()

Error de las predicciones de españa dividida por marca.

In [884]:
import numpy as np
import plotly.graph_objects as go

# Iterar sobre cada submarca
for subbrand in subbrands:
    # Filtrar las ventas reales para la submarca actual
    sales_actual_subbrand = sales_actual[sales_actual['SUBBRAND'] == subbrand]

    # Filtrar las predicciones para esta submarca
    datos_forecast_subbrand = datos_forecaste[datos_forecaste['SUBBRAND'] == subbrand]

    # Crear lista para almacenar los errores MAE
    mae_list = []
    forecast_names = []

    # Iterar sobre cada tipo de predicción en esta submarca
    for forecast_type in datos_forecast_subbrand['FORECAST'].unique():
        forecast_data = datos_forecast_subbrand[datos_forecast_subbrand['FORECAST'] == forecast_type]

        # Agrupar las predicciones por mes y submarca
        sales_forecast = forecast_data.groupby(['DATE', 'SUBBRAND'])['AMOUNT'].sum().reset_index()

        # Unir las ventas reales con las predicciones para esta submarca
        merged_data = pd.merge(sales_actual_subbrand, sales_forecast, on='DATE', suffixes=('_real', '_forecast'))

        # Calcular MAE
        mae = np.mean(np.abs(merged_data['AMOUNT_real'] - merged_data['AMOUNT_forecast']))
        mae_list.append(mae)

        # Añadir el nombre de la predicción
        forecast_names.append(f'{forecast_type}')

    # Crear el gráfico de barras para esta submarca
    fig_errors = go.Figure()

    # Añadir barras para el MAE
    fig_errors.add_trace(go.Bar(
        x=forecast_names,
        y=mae_list,
        name='MAE (Error Absoluto Medio)',
        marker_color='lightblue'
    ))

    # Configuración del gráfico de errores para esta submarca
    fig_errors.update_layout(
        title=f'Error MAE para {subbrand}',
        xaxis_title='Tipo de Predicción',
        yaxis_title='Error (MAE)',
        legend_title='Tipo de Error',
        template="plotly_white",
        xaxis=dict(tickangle=45)
    )

    # Mostrar el gráfico de errores para esta submarca
    fig_errors.show()


Comparación de ventas y predicciones de todas las marcas en españa juntas.

In [885]:
datos_españa = datos_final[datos_final['COUNTRY'] == 'Spain'].copy()

datos_actuals = datos_españa[datos_españa['SCENARIO'] == 'actual'].copy()
datos_forecasts = datos_españa[datos_españa['SCENARIO'] == 'AI_forecast'].copy()

datos_actuals['DATE'] = pd.to_datetime(datos_actuals[['YEAR', 'MONTH']].assign(DAY=1))
datos_forecasts['DATE'] = pd.to_datetime(datos_forecasts[['YEAR', 'MONTH']].assign(DAY=1))

sales_actual = datos_actuals.groupby('DATE')['AMOUNT'].sum().reset_index()

fig_sales = go.Figure()

fig_sales.add_trace(go.Scatter(
    x=sales_actual['DATE'],
    y=sales_actual['AMOUNT'],
    mode='lines+markers',
    name='Ventas Reales (Actual)',
    line=dict(color='blue')
))

# Añadir cada línea de predicción
for forecast_type in datos_forecasts['FORECAST'].unique():
    forecast_data = datos_forecasts[datos_forecasts['FORECAST'] == forecast_type]
    sales_forecast = forecast_data.groupby('DATE')['AMOUNT'].sum().reset_index()

    fig_sales.add_trace(go.Scatter(
        x=sales_forecast['DATE'],
        y=sales_forecast['AMOUNT'],
        mode='lines+markers',
        name=f'Predicción {forecast_type}',
        line=dict(dash='dash')
    ))

fig_sales.update_layout(
    title='Comparación de Ventas Reales vs. Predicciones España',
    xaxis_title='Fecha',
    yaxis_title='Cantidad Ventas',
    legend=dict(
        orientation="v",
        x=1.1,
        y=1,
        xanchor='left',
        yanchor='top'
    ),
    template="plotly_white"
)

fig_sales.show()

Aunque no se pueda ver los actuals se ve que los Forecast se han igualado mucho al borrar los valores repetidos.

Error de españa con todas las marcas juntas.

In [886]:
# Crear listas para almacenar los errores (MAE)
mae_list = []
forecast_names = []

# Iterar sobre cada tipo de predicción
for forecast_type in datos_forecasts['FORECAST'].unique():
    forecast_data = datos_forecasts[datos_forecasts['FORECAST'] == forecast_type]

    # Agrupar las predicciones por mes
    sales_forecast = forecast_data.groupby('DATE')['AMOUNT'].sum().reset_index()

    # Unir las ventas reales con las predicciones
    merged_data = pd.merge(sales_actual, sales_forecast, on='DATE', suffixes=('_real', '_forecast'))

    # Calcular MAE
    mae = np.mean(np.abs(merged_data['AMOUNT_real'] - merged_data['AMOUNT_forecast']))
    mae_list.append(mae)

    # Añadir el nombre de la predicción
    forecast_names.append(f'{forecast_type}')

# Crear el gráfico de barras para los errores (MAE) para toda España
fig_errors = go.Figure()

# Añadir barras para el MAE
fig_errors.add_trace(go.Bar(
    x=forecast_names,
    y=mae_list,
    name='MAE (Error Absoluto Medio)',
    marker_color='blue'
))

# Configuración del gráfico de errores
fig_errors.update_layout(
    title='Error MAE para Predicciones en España',
    xaxis_title='Tipo de Predicción',
    yaxis_title='Error (MAE)',
    legend_title='Tipo de Error',
    template="plotly_white",
    xaxis=dict(tickangle=45),
    legend=dict(
        orientation="v",  # Para una leyenda vertical
        x=1.1,  # Mover la leyenda a la derecha
        y=1,  # Posición de la leyenda (ajustar según sea necesario)
        xanchor='left',  # Ancla la leyenda a la izquierda de la posición 'x'
        yanchor='top'  # Ancla la leyenda a la parte superior
    )
)

# Mostrar el gráfico de errores para toda España
fig_errors.show()


Eliminando los datos repetidos observamos que el error de AI_PF, AI_P02F y AI_P03F se mantiene igual que antes, sin embargo,en el resto de Forecast se estaviliza el error y dan errores similares. Siendo mucho más pequeños, ya que están los datos más cercaanos a los actual, sin valores repetidos.